### DISCUSSION TITLE
Is it possible to add multiple covariance types？

### DISCUSSION THREAD

**Discussion by sdodingnan** [AlexanderFabisch/gmr#47] [Q&A]
Thank you for your excellent work，I want to know if there is a more flexible GMR, similar to sklearn, which can limit the covariance type to ‘shared’, ‘spherical’ or ‘diag'. Sometimes we don’t need a 'full' type of covariance matrix. I look forward to your reply.

**Comment by AlexanderFabisch** [comment]
Hi and thanks.

You can always use sklearn for training and then pass the covariances to gmr like in this example: https://github.com/AlexanderFabisch/gmr/blob/master/examples/plot_iris_from_sklearn.py

Would that work for you? I don't plan to copy the feature to gmr.

**Reply by advaitkambli** [reply]
No it does't work with  ‘shared’, ‘spherical’  an tied

**Reply by AlexanderFabisch** [reply]
sklearn knows [four types of covariances](https://scikit-learn.org/stable/modules/generated/sklearn.mixture.GaussianMixture.html) and for each type it is possible to convert them to the full covariance matrix that is required by gmr:

* `full`: each component has its own general covariance matrix. In this case you can directly pass `sklearn_gmm.covariances_` as the covariances argument to the GMM.
* `tied`: all components share the same general covariance matrix. The matrix has to be replicated `n_components` times, e.g., `np.array([gmm_sklearn.covariances_ for _ in range(gmm_sklearn.n_components)])`.
* `diag`: each component has its own diagonal covariance matrix. Each covariance has to be converted into a full covariance matrix, e.g., by `np.array([np.diag(c) for c in gmm_sklearn.covariances_])`.
* `spherical`: each component has its own single variance. You can convert these variances to covariances via `np.array([cov * np.eye(gmm_sklearn.n_components) for cov in gmm_sklearn.covariances_])`

**Comment by sdodingnan** [comment]
> 嗨，谢谢。
> 
> 您始终可以使用 sklearn 进行训练，然后将协方差传递给 gmr，如下例所示：[https ://github.com/AlexanderFabisch/gmr/blob/master/examples/plot_iris_from_sklearn.py](https://github.com/AlexanderFabisch/gmr/blob/master/examples/plot_iris_from_sklearn.py)
> 
> 这对你有用吗？我不打算将该功能复制到 gmr。

I got it, thanks, but whether the GMR can be used for prediction with the same input and output, such as time series anomaly detection tasks, add some noise to the input, and fix it through GMR, if it is normal data, it can be reconstructed well, and the prediction errors will be very large if it is abnormal

**Comment by AlexanderFabisch** [comment]
This could be possible. Pass the noisy data as `X` and the denoised data as `y` to the method `GMR.fit`. Let me know whether it works as you need it. I don't know. My guess is that you cannot really distinguish noise that is not normally distributed from normally distributed noise if you only have a few samples.
